homeassistant/components/mqtt_room/sensor.py

```diff
@@ -118,20 +130,23 @@ def message_received(msg):
 
             device = _parse_update_data(msg.topic, data)
             if device.get(CONF_DEVICE_ID) == self._device_id:
-                if self._distance is None or self._updated is None:
-                    update_state(**device)
-                else:
+                for idx, rec in enumerate(self._records):
+                    if rec.get(ATTR_DISTANCE) is None or rec.get(ATTR_UPDATED) is None:
+                        update_state(idx, **device)
+                        break
+
                     # update if:
                     # device is in the same room OR
                     # device is closer to another room OR
                     # last update from other room was too long ago
-                    timediff = dt.utcnow() - self._updated
+                    timediff = dt.utcnow() - rec.get(ATTR_UPDATED)
                     if (
-                        device.get(ATTR_ROOM) == self._state
-                        or device.get(ATTR_DISTANCE) < self._distance
+                        device.get(ATTR_ROOM) == rec.get(ATTR_ROOM)
+                        or device.get(ATTR_DISTANCE) < rec.get(ATTR_DISTANCE)
                         or timediff.total_seconds() >= self._timeout
                     ):
-                        update_state(**device)
+                        update_state(idx, **device)
+                        break
 
         return await mqtt.async_subscribe(
```

Comment:
(Y)

Reply:
Strictly this is not related to your PR, so you can leave it here. But the unsubscribe handler is returned while there is no return value supported here. Unsubscribe is only needed if the integration supports reloading, which it does not at the moment.

Reply:
E.g.
```python
class MqttRoomState(TypedDIct):
    """Hold the state of a room."

    room: str
    distance: float
    updated: datetime
```
The class will act as a dict and you can use `.update()` to update the current state. Also you will not need `_create_record()` any more as you can pass the inital values to the typed dict.

Reply:
I see. I think it might be a little more complicated than that: 

a) The source device dictionary also contains an device_id key which is discarded and not stored in the room state.
b) the updated attribute is computed and I would like to avoid having to compute it in multiple locations to pass to update

I.e. I won't be able to fully get rid of some wrapping update_state or similar function to call the TypedDict's update method. But let me take a stab at it nevertheless.

Reply:
May be you are right, I think it is possible to add the device_id, you can alway covert to `dict` then pop the device_id when needed, I leave that up to you.